docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,505 @@
+
+### User Stories
+
+#### Story 1 - Use ClusterClass to easily stamp clusters
+As an end user, I want to use one `ClusterClass` to create multiple managed cluster topologies of similar flavor.
+- Rather than recreating the KCP and MD objects for every cluster that needs to be provisioned, the end user can create a template once and reuse it to create multiple clusters with similar configurations.
+
+#### Story 2 - Easier UX for kubernetes version upgrades
+For an end user, the UX to update the kubernetes version of the control plane and worker nodes in the cluster should be easy.
+- Instead of individually modifying the KCP and each MachineDeployment, updating a single option should result in k8s version updates for all the CP and worker nodes.
+
+**Note**: In order to complete the user story for all the providers, some of the advanced features (such as Extensibility/Transformation) are required. However, getting this in place even only for a subset of providers allows us to build and test a big chunk of the entire machinery.
+
+#### Story 3 - Easier UX for scaling workers nodes
+As an end user, I want to be able to easily scale up/down the number of replicas for each set of worker nodes in the cluster.
+- Currently, (for a cluster with 3 machine deployments) this is possible by updating these three different objects representing the sets of worker nodes in the pool. An easier user experience would be to update a single object to enable the scaling of multiple sets of worker nodes.
+
+### Implementation Details/Notes/Constraints
+
+The following section provides details about the introduction of new types and modifications to existing types to implement the ClusterClass functionality.
+If instead you are eager to see an example of ClusterClass and how the Cluster object will look like, you can jump to the Behaviour paragraph.
+
+#### New API types
+##### ClusterClass
+This CRD is a collection of templates that describe the topology of one or more managed clusters.
+```golang
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=clusterclasses,shortName=cc,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+
+// ClusterClass is a template which can be used to create managed cluster topologies.
+type ClusterClass struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec ClusterClassSpec `json:"spec,omitempty"`
+}
+
+// ClusterClassSpec describes the desired state of the ClusterClass.
+type ClusterClassSpec struct {
+  // ControlPlane is a reference to a local struct that holds the details
+  // for provisioning the Control Plane for the Cluster.
+  ControlPlane LocalObjectTemplate `json:"controlPlane,omitempty"`
+
+  // Worker describes the worker nodes for the cluster.
+  // It is a collection of node types which can be used to create
+  // the worker nodes of the cluster.
+  // +optional
+  Worker WorkerClass `json:"worker,omitempty"`
```

Comment:
why not name this workerClass?

Reply:
I assume the struct was called WorkerClass here since it's part of ClusterClass, and Worker is used in the extension to the Cluster resource.

Reply:
The idea was to clearly demarcate the ClusterClassSpec into:
- Control Plane node definition
- Worker node definition

Adding the `Class` prefix would make it appear as `workerClass.Deployments[0].Class` which makes the naming scheme slightly confusing.

Reply:
Copying over a discussion from the Google doc which is about WorkerClass struct too



> Puja Abbassi
> 4:47 AM Jun 1
> Is this an "3 similar node pool each in a different AZ" use case only? Otherwise, I'm not sure I've seen end users scaling their node pools within a single cluster in the same way, especially when the node pools have different instance/machine sizes. 
> Maybe for multi-tenant clusters that have a node pool per team that could be a use case, but tbh I've not seen this use case a lot.
> I can however think of use cases where the same node pool in different clusters should be scaled similarly.
> Most probably as long as the WorkerClass can be used within and across clusters we can serve both types of use cases.
> 
> Sagar Muchhal
> 10:45 AM Jun 1
> The intent of the example was to showcase that currently to scale different sets of worker nodes, we need to update the MD objects representing those sets of worker nodes.
> Instead of going through multiple object updates, we can individually update the replica number for each of these worker sets under the `spec.Managed` field and thereby just having to update a single object once (with diff number of replicas for each set of worker nodes).
> Hopefully this resolves any confusion that the current wording may have caused.
> 
> Puja Abbassi
> 11:26 AM Jun 1
> I was just a bit confused, cause above there's a mention of WorkerClass and I thought that would be a separate object, but below I do not see it anymore, or rather workers are defined within a CC. The example you describe makes sense in the context of managing sets of workers within a single cluster.I was just thinking that there are similarly use cases for managing sets of workers across different clusters that have similar needs but might have differences on the configuration of the CC. That made me wonder if having worker class and cluster class separated could be valuable to create a more modular approach to managing clusters. Is that out of scope here?
> 
> Sagar Muchhal
> 11:57 AM Jun 1
> That is an interesting thought, of having separate class types to individually control the CP and worker nodes, and making the CC a collection of these classes. I would like to hear more from the community about what they think about this usecase.
> The naming `WorkerClass` was just to keep the objects aligned with the concept that they belong to. While outlining the proposal, this wasn't a usecase we designed for. Plus inlining the definition of the WorkerClass object makes it easier to define a CC.